### DISCUSSION TITLE
Plain HML + JS

### DISCUSSION THREAD

**Discussion by AndreiMotinga** [tailwindlabs/tailwindcss#5096] [Ideas]
Any plans to build components with plain HTML and JS, or at least something other than React or Vue? 

I use Rails in a standard MVC way. especially with new turbo js, looks like it's making a comeback for simple full-stack MVC apps, something doesn't need the full complexity of front-end library.  Don't you guys think you missing out a significant market by not including premade components for vanilla js?

**Comment by simonswiss** [comment]
Hey!

Assuming you're talking about Tailwind UI, the main reason we're not shipping HTML components with vanilla JS is this:

A truly complete, accessible implementation of UI patterns (like modals, dropdowns menus, ...) requires **a lot** of JavaScript. Robust keyboard navigation, correct ARIA attribute management, etc.

The last thing we want is to have basic JS support (open/close menu, "visual behaviour" stuff only) and have folks ship this to their production apps, assuming it's a finished product.

We _could_ provide that code, but I am not sure what a good "snippets ergonomic" would look like. Providing that code with each snippet would lead to mayhem, since folks would use multiple components the duplicated logic.

Provide code on a CDN file? This still makes a lot of assumptions on the users' markup to set up events listeners and wire DOM updates.

It's quite tricky to find a great way to "package" this. This hard problem is one of the main reasons that lead us to create Headless UI: an elegant way to abstract all the complexity of accessible / ARIA-compliant UI components, while retaining great "code snippet ergonomics".

We might explore other ways to package this logic – Alpine JS could potentially be a good fit for projects that don't need to go the "full JS framework" route.

Plain HTML + vanilla JS is much harder to elegantly package away.

Hopefully this answers your question!

**Reply by AndreiMotinga** [reply]
I tried once and struggled to get things working. 

Now I'm trying this again. looking into the alpine, so I want to try once more. 

But I can't help to think that: 

1. some people are definitely using it without front-end framework, so I'm pretty sure they simply manually recreate this. with alpine or just some js. There's gotta be an easier way than reinventing the wheel. 

2. I probably miss something here, but the most common things that come to mind are 
  - dropdown
  - tabs
  - modal
  
  this doesn't sound hard. 
    
  I assume this will do for reusing dropdown: https://alpinejs.dev/essentials/state#re-usable-data
  going to be trying this. 

  From the outside view, comparing this to something like bootstrap, I'm thinking bootstrap ships css and js. 
  if you dont' include js, these things won't work. Can Tailwind ship some js for these? 
  
  
  > The last thing we want is to have basic JS support (open/close menu, "visual behaviour" stuff only) and have folks ship this to their production apps, assuming it's a finished product.
  
  Well, isn't this what I'm going to end up with if I want to use this with rails without react? 
  
 You show me how it works(may work) and telling me that I may use alpine(but do it myself) - sorta sounds like you give me a foot and a gun :) ... recipe for disaster. 
 
 
 > We might explore other ways to package this logic – Alpine JS could potentially be a good fit for projects that don't need to go the "full JS framework" route.
 
 Yes, it would help a right now great deal. this won't work for everybody but will work for some/many. 
 I mean if I have to include apline in my project, similar to having to include bootstraip.js, then why not?

**Reply by simonswiss** [reply]
Hey!

> I probably miss something here, but the most common things that come to mind are
>
> dropdown
tabs
modal
this doesn't sound hard.

It's actually _really_ hard, if you want to implement a ARIA-compliant, fully accessible implementation of these.

Making tabs / dropdown visually work (click to open, click to close) is easy, but there's much more to it if you want a _proper_ implementation.

What I mean by that is supporting all the expected HTML attributes, keyboard navigation behaviour, etc:

Here's the example for a [Dropdown Menu](https://www.w3.org/TR/wai-aria-practices/#menubutton):

![image](https://user-images.githubusercontent.com/485747/127433914-2c07eee5-50d9-4da3-a119-47836111c02f.png)

![image](https://user-images.githubusercontent.com/485747/127433970-5a10fa2a-aa73-4572-804d-30c18a9e4c79.png)

All of the above is implemented in a solution like Headless UI.

Bootstrap does a really good job at packing a lot of this in their JS, as you say. I had a look at the Bootstrap 5 docs just then. You would still need to do some of the implementation yourself to reach that level of accessibility, as they mention in the docs:

![image](https://user-images.githubusercontent.com/485747/127436087-69bdc445-2b41-4ce0-a189-04f47952df79.png)

Bootstrap and Tailwind are different products though. For better or worse, Tailwind CSS is a CSS framework that makes as little assumption as possible on what you'll be building.

Bootstrap gives you pre-made building blocks (Navbar, carousel, etc), which come with batteries (JS) included.

Tailwind UI brings this layer of "opinion", but also still uses 100% Tailwind under the hood.

It's much harder to provide a JS implementation when you're not defining strict HTML markup patterns that you re-use everywhere, like Bootstrap components do.

I am not saying we are refusing to do it, just underlying that this is a harder problem than it appears on the surface. 

We used to ship with AlpineJS baked in, actually – but most users assumed that this meant the components were production-ready. Which - in terms of accessibility, they were not.

We take a11y very seriously - and try to make sure what we ship have been thoroughly vetted in that regard.

Hopefully, an Alpine JS version of Headless UI would be the solution for this situation for folks using Rails, PHP, HTML, anything which is not React/Vue.

We need to do some discovery work on that front before saying we're committing to doing it, because there might be huge blockers that we haven't uncovered yet to make it feasible the way we imagine it could be.

**Comment by ghost** [comment]
well then mention on the demo page that js is not provided/included because when I see "js required" seems like a helper text saying it will need js to work. but do not expect to not have js after purchase I know it is hidden in terms or faqs but do us a favour and mention it where you say js required some additional text that you are actually not providing js even if I make a purchase
